### ISSUE TITLE
AWS CloudTrail integration improvements

### ISSUE THREAD

**Issue by JPLachance** [wazuh/wazuh#510]
Greetings,

I enabled the AWS CloudTrail integration and I've let it run for several hours but I got no results. My AWS access key was working and from the manager, I was seeing the Python process using all its available resources.

Then, I decided to [read the code](https://github.com/wazuh/wazuh/blob/master/wodles/aws/aws.py#L124), and I saw my issue:
```python
for bucket_file in my_bucket.objects.all():
    # Do stuff...
```

The AWS CloudTrail integration is currently a loop going through the entire CloudTrail S3 bucket. In our case, we have up to 1 year of data in that bucket. I tested the loop locally and it is almost infinite 😄

We already have one trail enabled and enabling a second trail dedicated to Wazuh would not be free. The second Trail in CloudTrail can be really expensive!

So, instead of reading all the files from S3, we could create an AWS Lambda function that gets triggered by the `s3:ObjectCreated:*` and then sends events to Wazuh. AWS provided [a documentation on the subject](https://docs.aws.amazon.com/lambda/latest/dg/with-cloudtrail.html). Since it requires to create multiple resources in AWS, the Wazuh documentation could provide a [Terraform module](https://www.terraform.io/docs/modules/index.html) to automate the setup. That module could [create the Lambda function](https://www.terraform.io/docs/providers/aws/r/lambda_function.html) and all the other resources to make that work.

Or, another option is to read only the files in S3 from the last couple of hours (or from the last `<interval>10m</interval>` setting).

What do you think?

Best regards,

Jean-Philippe Lachance

**Comment by UranusBytes** [comment]
I'm much in the same boat as you.  I was thinking of spinning up Wazuh quickly for some CloudTrail monitoring.  In my case, I have multiple S3 buckets that store CloudTrail for multiple AWS accounts, with multiple years of logs stored.  The script will burst into flames trying to list all the objects in those buckets...

While I agree that a SNS push/event based system or SQS pull/queue based system would be more elegant, I'm thinking a first step would be a simple "only grab new objects since X" would be easier to both implement, and reduce dependencies upon other AWS services.

Based upon this, what I'm thinking of trying to attempt would be something along the lines of:

1. All the configuration of a prefix in the wodle (since CloudTrail allows a key prefix to be defined; allows for those circumstances where other stuff stored in S3 bucket than just CloudTrail)
2. Allow to define specific AWS account numbers (for circumstances where only want to monitor a subset of AWS account CloudTrail stored in S3 bucket)
3. Only pull new logs since last execution

Other things I'm thinking about:
- Some amount of pruning of the log_progress table (maybe all entries older than the last X days of processed logs???)
- Grabbing logs from multiple S3 buckets (this is a problem I'm facing, but I may punt)

I'm specifically ignoring:
- Only parsing CloudTrail logs from specific regions
- Dealing with encrypted logs (not my problem today)
- Integrity check of CloudTrail logs, assuming digest is enabled (
- Use IAM instance profile/roles (would much prefer over embedded keys, but I can punt on this)

Assuming I get some spare cycles over the next week or two, I'll take a stab at this...

**Comment by UranusBytes** [comment]
I've got a working version of this committed into this fork branch - https://github.com/UranusBytes/wazuh/tree/510

It isn't a completely transparent for existing/legacy configs (it doesn't yet migrate the old sqlite DB to new format), so it would re-process existing logs. The skinny on using it to define multiple "cloudtrail" nodes in the aws-cloudtrail wodle config as follows:
```

  <wodle name="aws-cloudtrail">
    <disabled>no</disabled>
    <bucket>bucket-name</bucket> <!-- DEPRECATED, but still supported -->
    <access_key></access_key> <!-- DEPRECATED, but still supported -->
    <secret_key></secret_key> <!-- DEPRECATED, but still supported -->
    <remove_from_bucket>no</remove_from_bucket> <!-- DEPRECATED, but still supported -->
    <interval>10m</interval>
    <run_on_start>no</run_on_start>
    <error_not_skip>yes/no<error_not_skip> <!-- OPTIONAL: On failure to parse log, error/exit wodle instead of skip (default) -->
    <cloudtrail>
      <bucket></bucket> <!-- REQUIRED: Name of S3 bucket -->
      <aws_account_id></aws_account_id> <!-- REQUIRED: The   -->
      <aws_account_alias></aws_account_alias> <!-- OPTIONAL: User friendly alias for AWS account   -->
      <access_key></access_key> <!-- OPTIONAL: IAM access key for get from s3 -->
      <secret_key></secret_key> <!-- OPTIONAL: IAM secret key for get from s3 -->
      <aws_profile></aws_profile> <!-- OPTIONAL: profile from cred file for get from s3 -->
      <iam_role_arn></iam_role_arn> <!-- OPTIONAL: IAM role for get from s3 -->
      <prefix></prefix> <!-- OPTIONAL: a prefix defined for CloudTrail -->
      <only_logs_after></only_logs_after> <!-- OPTIONAL: Date in YYYY-MMM-DD format to only parse logs after -->
      <region></region> <!-- OPTIONAL: comma-delimited list of regions to process -->
      <remove_from_bucket>no</remove_from_bucket> <!-- OPTIONAL: delete the logs from S3 -->
    </cloudtrail> 

```
Before I can do a pull request, I still need to:
- Test out using IAM role for access to S3
- Write up better documentation
- Finish the 'migration' function for legacy installs

**Comment by mgmacias95** [comment]
Hello all,

I'm closing this issue since this is already implemented in #1105 and #913 .

Best regards,
Marta
